cli_tools/gce_windows_upgrade/upgrader/upgrader.go

```diff
@@ -0,0 +1,528 @@
+		"\n" +
+		"When upgrading failed but you didn't enable auto-rollback, or auto-rollback failed, or " +
+		"upgrading succeeded but you still need to rollback for any reason, " +
+		"please manually rollback by following steps:\n" +
+		"1. Detach the new OS disk from the instance and delete it.\n" +
+		"2. Attach the old OS disk as boot disk.\n" +
+		"3. Detach the install media disk from the instance and delete it.\n" +
+		"4. Delete 'windows-startup-script-url' from the instance's metadata if there isn't an original value. " +
+		"If there is an original value, restore it. The original value is backed up as metadata 'windows-startup-script-url-backup'.\n" +
+		"\n" +
+		"Once you verified the upgrading succeeded and decided to never rollback, you can:\n" +
+		"1. Delete the original OS disk.\n" +
+		"2. Delete the machine image.\n" +
+		"3. Delete the snapshot.\n" +
+		"\n"
+)
+
+var (
+	supportedSourceOSVersions = map[string]string{versionWindows2008r2: versionWindows2012r2}
+	supportedTargetOSVersions = reverseMap(supportedSourceOSVersions)
+
+	upgradeScriptName        = map[string]string{versionWindows2008r2: "upgrade_script_2008r2_to_2012r2.ps1"}
+	upgradeWorkflowPath      = map[string]string{versionWindows2008r2: "daisy_workflows/windows_upgrade/windows_upgrade_2008r2_to_2012r2.wf.json"}
+	retryUpgradeWorkflowPath = map[string]string{versionWindows2008r2: "daisy_workflows/windows_upgrade/windows_upgrade_2008r2_to_2012r2_retry.wf.json"}
+
+	expectedLicense = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2008-r2-dc"}
+	appendLicense   = map[string]string{versionWindows2008r2: "projects/windows-cloud/global/licenses/windows-server-2012-r2-dc-in-place-upgrade"}
+
+	instanceURLRgx = regexp.MustCompile(fmt.Sprintf(`^(projects/(?P<project>%[1]s)/)?zones/(?P<zone>%[2]s)/instances/(?P<instance>%[2]s)$`, projectRgxStr, rfc1035))
+
+	windowsStartupScriptURLBackup       *string
+	windowsStartupScriptURLBackupExists bool
+
+	computeClient daisyCompute.Client
+)
+
+type validatedParams struct {
+	project          string
+	zone             string
+	instanceName     string
+	osDisk           string
+	osDiskType       string
+	osDiskDeviceName string
+	osDiskAutoDelete bool
+}
+
+// UpgradeParams contains user input params and validated (transformed) params
+type UpgradeParams struct {
+	ClientID               string
+	InstanceURI            string
+	SkipMachineImageBackup bool
+	AutoRollback           bool
+	SourceOS               string
+	TargetOS               string
+	ProjectPtr             *string
+	Timeout                string
+	ScratchBucketGcsPath   string
+	Oauth                  string
+	Ce                     string
+	GcsLogsDisabled        bool
+	CloudLogsDisabled      bool
+	StdoutLogsDisabled     bool
+	CurrentExecutablePath  string
+
+	validatedParams
+}
+
+// Run runs upgrade workflow.
+func Run(upgradeParams *UpgradeParams) (*daisy.Workflow, error) {
```

Comment:
Instead of params you could just make this a method of the Upgrade struct, passing a large params struct around seems odd.